Sources/SourceKitLSP/SourceKitServer.swift

```diff
@@ -855,28 +855,49 @@ extension SourceKitServer {
     languageService.documentColor(req)
   }
 
+  private func supportsSemanticTokens(in uri: DocumentURI) -> Bool {
+    guard let workspace = workspace,
+          let snapshot = workspace.documentManager.latestSnapshot(uri) else {
+      return false
+    }
+    let language = snapshot.document.language
+    return workspace.capabilityRegistry.hasSemanticTokensRegistration(for: language)
+  }
+
   func documentSemanticTokens(
     _ req: Request<DocumentSemanticTokensRequest>,
     workspace: Workspace,
     languageService: ToolchainLanguageServer
   ) {
-    languageService.documentSemanticTokens(req)
+    if supportsSemanticTokens(in: req.params.textDocument.uri) {
```

Comment:
This was (briefly) discussed in https://github.com/apple/sourcekit-lsp/pull/414#discussion_r666021462.

The idea is to make sure semantic tokens aren't provided in a Swift file if `clangd`'s semantic tokens provider (with its own legend, etc.) was already registered, which IIUC happens if the user opens an (Obj)C(++) file first. Although it would be nice to support semantic tokens in both Swift and (Obj)C(++) files simultaneously, coordinating the (possibly different) legends and recoding the tokens if needed would add quite a bit of complexity that I would consider out-of-scope for this PR.

Reply:
We currently only support dynamic registration for semantic highlighting - we don't register static capabilities [here](https://github.com/apple/sourcekit-lsp/blob/7c9ccf26d5f758a23679ca04a2ddf6238e94af3a/Sources/SourceKitLSP/SourceKitServer.swift#L527) and we dynamically register support [here](https://github.com/apple/sourcekit-lsp/blob/7c9ccf26d5f758a23679ca04a2ddf6238e94af3a/Sources/SourceKitLSP/SourceKitServer.swift#L573), so this is unnecessary. If we wanted to support both clangd and a Swift/sourcekit-lsp legend with static registration then we'd need more logic like you mentioned.

Reply:
Ah, so the dynamic registration mechanism already takes care of ensuring that we only register a semantic tokens provider for one language service and we can remove this extra check? That's nice.